add docs for enforce mpp (#5811)
pingcap/docs: tiflash/use-tiflash.md

## Use the MPP mode

TiFlash supports using the MPP mode to execute queries, which introduces cross-node data exchange (data shuffle process) into the computation. TiDB automatically determines whether to choose the MPP mode using the optimizer. You can change the selection strategy by modifying the values of [`tidb_allow_mpp`](/system-variables.md#tidb_allow_mpp-new-in-v50) and [`tidb_enforce_mpp`](/system-variables.md#tidb_enforce_mpp-new-in-v51).

### Control whether to choose the MPP mode

> `tidb_allow_mpp` controls whether TiDB can select MPP mode to execute queries. `tidb_enforce_mpp` controls whether the optimizer cost estimate is ignored and the MPP mode of TiFlash is forced to execute queries.

Reviewer: ```suggestion
The `tidb_allow_mpp` variable controls whether TiDB can select the MPP mode to execute queries. The `tidb_enforce_mpp` variable controls whether the optimizer's cost estimation is ignored and the MPP mode of TiFlash is forcibly used to execute queries.
```

The results corresponding to all values of these two variables are as follows:

|                        | tidb_allow_mpp=off | tidb_allow_mpp=on (by default)              |
| ---------------------- | -------------------- | -------------------------------- |

> | tidb_enforce_mpp=off (by default) | MPP mode is not used. | The optimizer selects the MPP mode based on cost estimation. (by default)|
> | tidb_enforce_mpp=on  | MPP mode is not used.   | TiDB ignores the cost estimate and chooses MPP mode.      |

Reviewer: ```suggestion
| tidb_enforce_mpp=off (by default) | The MPP mode is not used. | The optimizer selects the MPP mode based on cost estimation. (by default)|
| tidb_enforce_mpp=on  | MPP mode is not used.   | TiDB ignores the cost estimation and chooses the MPP mode.      |
```

For example, if you do not want to use MPP mode, you can execute the following statements:

{{< copyable "sql" >}}

```sql
set @@session.tidb_allow_mpp=1;
set @@session.tidb_enforce_mpp=0;
```

> If you want to use the cost estimation of TiDB optimizer to intelligently choose whether to use MPP mode (by default), you can execute the following statements:

Reviewer: ```suggestion
If you want TiDB to use the optimizer's cost estimation to intelligently choose whether to use the MPP mode (by default), you can execute the following statements:
```

> If you want to intelligently choose whether to use the MPP mode (by default) using the cost estimation of TiDB optimizer, you can execute the following statement:

Reviewer 2: This condition seems too long to read. How about replacing `using the cost est of TiDB optimizer` by 'by TiDB optimizer'? That makes sense to readers

{{< copyable "sql" >}}

```sql
set @@session.tidb_allow_mpp=1;
set @@session.tidb_enforce_mpp=0;
```

> If you want TiDB to ignore the optimizer's cost estimates and force to choose the MPP mode, you can execute the following statement:

Reviewer 2: and to force?

Reviewer 2: statments?

> If you want TiDB to ignore the optimizer's cost estimations and force to choose the MPP mode, you can execute the following statements:

Reviewer: ```suggestion
If you want TiDB to ignore the optimizer's cost estimation and to forcibly choose the MPP mode, you can execute the following statements:
```

{{< copyable "sql" >}}

```sql
set @@session.tidb_allow_mpp=1;
set @@session.tidb_enforce_mpp=1;
```

> The initial value of the session variable `tidb_enforce_mpp` is equal to the[`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration item value of this tidb-server instance (which is `false` by default). If you want to make several tidb servers only serve analytical quries in a TiDB cluster and to force them to use MPP mode, you can change their [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration value to `true`.

Reviewer 3: ```suggestion
The initial value of the session variable `tidb_enforce_mpp` is equal to the [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration item value of this tidb-server instance (which is `false` by default). If you want to make several tidb servers only serve analytical quries in a TiDB cluster and to force them to use MPP mode, you can change their [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration value to `true`.
```

> The initial value of the Session variable `tidb_enforce_mpp` is equal to the[`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration item value of this tidb-server instance (which is `false` by default). In a TiDB cluster, if several tidb-server instances only perform analytical queries, to ensure that they can choose MPP mode, you can change their [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration value to `true`.

Reviewer 2: is "Session" need to capitalize the first letter?

Reviewer 2: When I first read the sentence "In a TIDB cluster ...", I'm quite confused.  I think you want to express "If you want to make several tidb servers only serve analytical quries in the cluster and to force them to use MPP mode," I think this way is clearer.

> The initial value of the session variable `tidb_enforce_mpp` is equal to the [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration item value of this tidb-server instance (which is `false` by default). If you want to make several tidb servers only serve analytical quries in a TiDB cluster and to force them to use MPP mode, you can change their [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration value to `true`.

Reviewer: ```suggestion
The initial value of the `tidb_enforce_mpp` session variable is equal to the [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration value of this tidb-server instance (which is `false` by default). If multiple tidb-server instances in a TiDB cluster only perform analytical queries and you want to make sure that the MPP mode is used on these instances, you can change their [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration values to `true`.
```

> **Note:**
>

> > When `tidb_enforce_mpp=1` takes effect, TiDB optimizer will ignore the cost estimation to choose MPP mode. However, TiDB will not select MPP mode if other factors block MPP mode, such as no TiFlash replica, the replication of TiFlash replicas is not completed, and statements contain operators or functions that are not supported by MPP mode.

Reviewer: ```suggestion
> When `tidb_enforce_mpp=1` takes effect, the TiDB optimizer will ignore the cost estimation to choose the MPP mode. However, if other factors block the MPP mode, TiDB will not select the MPP mode. These factors include the absence of TiFlash replica, unfinished replication of TiFlash replicas, and statements containing operators or functions that are not supported by MPP mode.
```

> > When `tidb_enforce_mpp=1` takes effect, the TiDB optimizer will ignore the cost estimation to choose MPP mode. However, TiDB will not choose the MPP mode if other factors that do not support the MPP mode occur, such as no TiFlash copy, the replication of TiFlash copies is not completed, and statements contain operators or functions that are not supported by the MPP mode.

Reviewer 2: It's not natural to use "copy" to describe the follower role in a cluster. The most often used term is "replica". "if other factors that do not support the MPP mode occur" is a little awkward. If I wrote it, I would say "However, there are still some cases that block MPP mode, such as missing TiFlash replica for some tables ..."

> 

> > If the TiDB optimizer cannot select MPP mode due to reasons other than cost estimation, when you use the `EXPLAIN` statement to view the execution plan, a warning is returned to explain the reason, for example:

Reviewer 2: sorry to be frank, this sentence is quite messy, so I think we have to reorganize the sentence. How about "When TiDB optimizer choose non-MPP plan unexpectly, we can use `explain` statement to check out the plan and the warnings which explain why it doesn't apply MPP mode". For example

## Use the MPP mode

> TiFlash supports using the MPP mode to execute queries, which introduces cross-node data exchange (data shuffle process) into the computation. TiDB automatically determines whether to select the MPP mode using the optimizer. You can change the selection strategy by modifying the values of [`tidb_allow_mpp`](/system-variables.md#tidb_allow_mpp-new-in-v50) and [`tidb_enforce_mpp`](/system-variables.md#tidb_enforce_mpp-new-in-v51).

Reviewer: ```suggestion
TiFlash supports using the MPP mode to execute queries, which introduces cross-node data exchange (data shuffle process) into the computation. TiDB automatically determines whether to select the MPP mode using the optimizer's cost estimation. You can change the selection strategy by modifying the values of [`tidb_allow_mpp`](/system-variables.md#tidb_allow_mpp-new-in-v50) and [`tidb_enforce_mpp`](/system-variables.md#tidb_enforce_mpp-new-in-v51).
```

### Control whether to select the MPP mode

The `tidb_allow_mpp` variable controls whether TiDB can select the MPP mode to execute queries. The `tidb_enforce_mpp` variable controls whether the optimizer's cost estimation is ignored and the MPP mode of TiFlash is forcibly used to execute queries.

The results corresponding to all values of these two variables are as follows:

|                        | tidb_allow_mpp=off | tidb_allow_mpp=on (by default)              |
| ---------------------- | -------------------- | -------------------------------- |
| tidb_enforce_mpp=off (by default) | The MPP mode is not used. | The optimizer selects the MPP mode based on cost estimation. (by default)|
| tidb_enforce_mpp=on  | The MPP mode is not used.   | TiDB ignores the cost estimation and selects the MPP mode.      |

For example, if you do not want to use the MPP mode, you can execute the following statements:

{{< copyable "sql" >}}

```sql
set @@session.tidb_allow_mpp=1;
set @@session.tidb_enforce_mpp=0;
```

> If you want TiDB to use the optimizer's cost estimation to intelligently select whether to use the MPP mode (by default), you can execute the following statements:

Reviewer: @windtalker PTAL thanks!
```suggestion
If you want TiDB to consider the optimizer's cost estimation to intelligently select whether to use the MPP mode (by default), you can execute the following statements:
```

Reviewer 2: How about "If you want TiDB's cost-based optimizer to automatically decide whether to use MPP mode.". Cost-based opimizer is a more common terms and makes sense to most readers.

Reviewer: committed.

set @@session.tidb_allow_mpp=1;
set @@session.tidb_enforce_mpp=0;
```

If you want TiDB to ignore the optimizer's cost estimation and to forcibly select the MPP mode, you can execute the following statements:

{{< copyable "sql" >}}

```sql
set @@session.tidb_allow_mpp=1;
set @@session.tidb_enforce_mpp=1;
```

The initial value of the `tidb_enforce_mpp` session variable is equal to the [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration value of this tidb-server instance (which is `false` by default). If multiple tidb-server instances in a TiDB cluster only perform analytical queries and you want to make sure that the MPP mode is used on these instances, you can change their [`enforce-mpp`](/tidb-configuration-file.md#enforce-mpp) configuration values to `true`.

> **Note:**
>
> When `tidb_enforce_mpp=1` takes effect, the TiDB optimizer will ignore the cost estimation to choose the MPP mode. However, if other factors block the MPP mode, TiDB will not select the MPP mode. These factors include the absence of TiFlash replica, unfinished replication of TiFlash replicas, and statements containing operators or functions that are not supported by the MPP mode.
> 

> > If TiDB optimizer cannot select the MPP mode due to reasons other than cost estimation, when you use the `EXPLAIN` statement to view the execution plan, a warning is returned to explain the reason. For example:

Reviewer 2: "to view" is unnatural. line 329 in this file uses 'check'. That makes more sense, such as "check out something".

Reviewer: ```suggestion
> If TiDB optimizer cannot select the MPP mode due to reasons other than cost estimation, when you use the `EXPLAIN` statement to check out the execution plan, a warning is returned to explain the reason. For example:
```
